mantidproject/mantid
# Fix access violation in MonteCarloAbsorption

**Description of work.**

The unit test suite MonteCarloAbsorptionTest has occasionally been failing on Windows with an access violation on Jenkins. I've been able to track this down to the test Lambda_StepSize_Two_Linear_Interpolation which was failing

I can reproduce this locally (even in a debug build) by modifying the test to repeat 1000 times and it eventually fails. Although the access violation occurs in a couple of different places:
a) the y() or e() vector in the 0th histogram is sometimes empty once the MonteCarloAbsorption algorithm completes which causes some of the test asserts to fall over because they assume y(0).front() or e(0).front is valid
b) sometimes the code in MonteCarloAbsorption.cpp itself has an access violation when yIndexOfX on ~L404 returns an index out of range of the histogram

Both problems happen in the multi-threaded code in MonteCarloAbsorption::doSimulation and it seems to be the interaction between the call to MatrixWorkspace::yIndexOfX and the code that updates a histogram with interpolated results via MatrixWorkspace::setHistogram. Something in these two functions isn't threadsafe. I found that yIndexOfX calls IsHistogramData() which checks the 0th histogram in the workspace even if the MonteCarloAbsorption code is processing the 1st, 2nd etc workspace. I think if the 0th histogram is in the process of being updated by another thread then yIndexOfX can fail or else the setHistogram call fails.

I have done two changes:
- only call yIndexOfX once (there was a redundant 2nd call)
- I have modified yIndexOfX so it takes an additional histogram index parameter. This allows me to keep each MonteCarloAbsorption thread interacting with a single histogram only. I have defaulted the new parameter to zero so other
code behaves as before

Don't know why this only happened on Windows

**To test:**

The problem isn't v easy to reproduce so mainly a case of just doing a code review. 

You can encourage it to happen by changing the code in MonteCarloAbsorptionTest to run the test 1000 times as follow (but maybe changing the code isn't normally within scope of PR testing so this is optional). This change drops the test_ prefix off the actual test method and adds a new test_ method that loops 1000 times:

```
  void test_Lambda_StepSize_Two_Linear_Interpolation() {
    for (int i = 0; i < 1000; i++) {
      Lambda_StepSize_Two_Linear_Interpolation();
    }
  }

  void Lambda_StepSize_Two_Linear_Interpolation() {
etc....
```

If you do this change then you should see an access violation before these code changes and no access violation after. May need to run the test a few times to see the access violation even with the for loop in place. I suggest you do a Release build because some of the tests in this suite are significantly slower in a Debug build.

Fixes #29624 .

*This does not require release notes* because **it has no functional impact**

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
